distributed/comm/asyncio_tcp.py

```diff
@@ -0,0 +1,781 @@
+            self._servers = [server]
+
+    def stop(self):
+        # Stop listening
+        for server in self._servers:
+            server.close()
+
+    def get_host_port(self):
+        """
+        The listening address as a (host, port) tuple.
+        """
+        if self.bound_address is None:
+
+            def get_socket(server):
+                for family in [socket.AF_INET, socket.AF_INET6]:
+                    for sock in server.sockets:
+                        if sock.family == family:
+                            return sock
+                raise RuntimeError("No active INET socket found?")
+
+            sock = get_socket(self._servers[0])
+            self.bound_address = sock.getsockname()[:2]
+        return self.bound_address
+
+    @property
+    def listen_address(self):
+        """
+        The listening address as a string.
+        """
+        return self.prefix + unparse_host_port(*self.get_host_port())
+
+    @property
+    def contact_address(self):
+        """
+        The contact address as a string.
+        """
+        host, port = self.get_host_port()
+        host = ensure_concrete_host(host, default_host=self.default_host)
+        return self.prefix + unparse_host_port(host, port)
+
+
+class TLSListener(TCPListener):
+    prefix = "tls://"
+    comm_class = TLS
+
+    def _get_extra_kwargs(self, address, **kwargs):
+        ctx = _expect_tls_context(kwargs)
+        return {"ssl": ctx}
+
+
+class TCPBackend(Backend):
+    _connector_class = TCPConnector
+    _listener_class = TCPListener
+
+    def get_connector(self):
+        return self._connector_class()
+
+    def get_listener(self, loc, handle_comm, deserialize, **connection_args):
+        return self._listener_class(loc, handle_comm, deserialize, **connection_args)
+
+    def get_address_host(self, loc):
+        return parse_host_port(loc)[0]
+
+    def get_address_host_port(self, loc):
+        return parse_host_port(loc)
+
+    def resolve_address(self, loc):
+        host, port = parse_host_port(loc)
+        return unparse_host_port(ensure_ip(host), port)
+
+    def get_local_address_for(self, loc):
+        host, port = parse_host_port(loc)
+        host = ensure_ip(host)
+        if ":" in host:
+            local_host = get_ipv6(host)
+        else:
+            local_host = get_ip(host)
+        return unparse_host_port(local_host, None)
+
+
+class TLSBackend(TCPBackend):
+    _connector_class = TLSConnector
+    _listener_class = TLSListener
+
+
+_LARGE_BUF_LIMIT = 2048
+
+
+class _ZeroCopyWriter:
+    """The builtin socket transport in asyncio makes a bunch of copies, which
+    can make sending large amounts of data much slower. This hacks around that.
+    Note that this workaround isn't used on windows or uvloop"""
+
+    def __init__(self, transport):
+        self.transport = transport
+        self._buffers = collections.deque()
+        self._first_pos = 0
+        self._size = 0
+
+    def _buffer_append(self, data):
```

Comment:
A previous commit implemented a more complicated buffer structure (where small consecutive frames were merged, and larger frames were kept zero-copy), but this didn't prove to matter in real life. Most messages seem to be `[small header, large frame, large frame, large frame, ...]`, so consecutive small frames are rare except for small messages (which we already merge into a single buffer earlier).

Reply:
Yeah, this is no longer true. For a comm-only benchmark (where the number of frames is low and each frame is large) the simpler buffer structure is much faster. But currently our serialization protocol can result in 1000s of tiny frames, which benefit from a more complicated buffer structure.